src/transports/tcp.rs

```diff
@@ -521,21 +559,48 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
 
         let result = try_write();
 
-        // after try_write we need to reset the timeout on the stream
-        endpoint.stream.set_write_timeout(None).map_err(|e| {
-            PeerNetError::SendError.error("error reset write timeout", Some(e.to_string()))
-        })?;
+        // after write we need to reset the timeout on the stream
+        endpoint
+            .stream
+            .set_write_timeout(Some(endpoint.config.write_timeout))
+            .map_err(|e| {
+                PeerNetError::SendError.error("error reset write timeout", Some(e.to_string()))
+            })?;
 
         result
     }
 
     fn receive(endpoint: &mut Self::Endpoint) -> PeerNetResult<Vec<u8>> {
         //TODO: Config one
+
+        let start_time = Instant::now();
         let mut len_bytes = vec![0u8; 4];
-        endpoint
-            .stream
-            .read_exact(&mut len_bytes)
-            .map_err(|err| TcpError::ConnectionError.wrap().new("recv len", err, None))?;
+
+        // read the length of the message
+        let mut len_read = 0;
+        while len_read < len_bytes.len() {
```

Comment:
move the logic to a `read_exact_timeout(stream: TcpStream, timeout: MassaTime, buffer: &mut[u8]) -> Result<(), Error>` function that can be called from here

Reply:
done